File: src/cmds/pkg.ts
import { run } from '../core/runner';
import { stripAnsi } from '../core/utils';
import { HandlerResult } from './git';

// Non-Node package managers and codegen: pip, uv, bundle, prisma, gem.
// Installs collapse to "N installed" + any errors; listings to counts.

export function handlePip(args: string[], ultra: boolean): HandlerResult {
  const sub = args[0] || '';
  const bin = 'pip';
  const result = run(bin, args);
  const raw = result.stdout + (result.stderr ? `\n${result.stderr}` : '');
  let filtered = '';
  try {
    if (sub === 'install') filtered = summarizePipInstall(raw, ultra);
    else if (sub === 'list' || sub === 'freeze') filtered = summarizeList(raw, ultra, 'packages');
    else filtered = stripAnsi(raw).trim(); // pip show/uninstall/…: full output
  } catch {
    filtered = raw;
  }
  return result0(filtered, result, `pip ${sub || 'cmd'}`, raw);
}

export function handleUv(args: string[], ultra: boolean): HandlerResult {
  const sub = args[0] || '';
  const result = run('uv', args);
  const raw = result.stdout + (result.stderr ? `\n${result.stderr}` : '');
  let filtered = '';
  try {
    if (sub === 'sync' || sub === 'pip' || sub === 'add' || sub === 'install') {
      filtered = summarizeUvSync(raw, ultra, result.exitCode);
    } else {
      // uv run <cmd> and everything else: pass the wrapped output through in full.
      filtered = stripAnsi(raw).trim();
    }
  } catch {
    filtered = raw;
  }
  return result0(filtered, result, `uv ${sub || 'cmd'}`, raw);
}

export function handleBundle(args: string[], ultra: boolean): HandlerResult {
  const sub = args[0] || 'install';
  const result = run('bundle', args);
  const raw = result.stdout + (result.stderr ? `\n${result.stderr}` : '');
  let filtered = '';
  try {
    const m = /Bundle complete!\s+(\d+)\s+Gemfile dependencies,\s+(\d+)\s+gems now installed/.exec(stripAnsi(raw));
    if (m) filtered = ultra ? `✓${m[2]}gems` : `✓ Bundle complete: ${m[2]} gems (${m[1]} deps)`;
    else if (result.exitCode !== 0) filtered = errorTail(raw);
    else filtered = stripAnsi(raw).trim(); // e.g. `bundle exec <cmd>`: pass wrapped output through
  } catch {
    filtered = raw;
  }
  return result0(filtered, result, `bundle ${sub}`, raw);
}

export function handlePrisma(args: string[], ultra: boolean): HandlerResult {
  const sub = args[0] || '';
  const result = run('prisma', args);
  const raw = result.stdout + (result.stderr ? `\n${result.stderr}` : '');
  let filtered = '';
  try {
    const clean = stripAnsi(raw);
    const gen = /Generated Prisma Client\s+\(([^)]+)\)/.exec(clean);
    const migrate = /(Your database is now in sync|already in sync|migration.+applied)/i.exec(clean);
    if (gen) filtered = ultra ? '✓gen' : `✓ Generated Prisma Client (${gen[1]})`;
    else if (migrate) filtered = ultra ? '✓migrate' : `✓ ${migrate[1]}`;
    else if (result.exitCode !== 0) filtered = errorTail(raw);
    else filtered = stripAnsi(raw).trim(); // prisma db pull/studio/…: full output
  } catch {
    filtered = raw;
  }
  return result0(filtered, result, `prisma ${sub || 'cmd'}`, raw);
}

export function handleGem(args: string[], ultra: boolean): HandlerResult {
  const sub = args[0] || '';
  const result = run('gem', args);
  const raw = result.stdout + (result.stderr ? `\n${result.stderr}` : '');
  let filtered = '';
  try {
    const m = /(\d+)\s+gems? installed/.exec(stripAnsi(raw));
    if (m) filtered = ultra ? `✓${m[1]}gems` : `✓ ${m[1]} gems installed`;
    else if (sub === 'list') filtered = summarizeList(raw, ultra, 'gems');
    else if (result.exitCode !== 0) filtered = errorTail(raw);
    else filtered = stripAnsi(raw).trim(); // other gem subcommands: full output
  } catch {
    filtered = raw;
  }
  return result0(filtered, result, `gem ${sub || 'cmd'}`, raw);
}

function summarizePipInstall(raw: string, ultra: boolean): string {
  const clean = stripAnsi(raw);
  const errors = clean.split('\n').filter((l) => /^ERROR:/.test(l));
  if (errors.length > 0) return ultra ? `✗${errors.length}err` : `✗ Install failed\n${errors.slice(0, 5).join('\n')}`;
  const m = /Successfully installed\s+(.+)/.exec(clean);
  if (m) {
    const pkgs = m[1].trim().split(/\s+/);
    if (ultra) return `✓${pkgs.length}pkg`;
    return `✓ Installed ${pkgs.length} package${pkgs.length === 1 ? '' : 's'}: ${pkgs.slice(0, 10).join(', ')}`;
  }
  if (/already satisfied/i.test(clean)) return ultra ? '✓cached' : '✓ requirements already satisfied';
  return ultra ? '✓' : '✓ ok';
}

function summarizeUvSync(raw: string, ultra: boolean, exitCode: number): string {
  const clean = stripAnsi(raw);
  const installed = (clean.match(/^\s*[+]\s+\S+/gm) || []).length;
  const removed = (clean.match(/^\s*[-]\s+\S+/gm) || []).length;
  if (exitCode !== 0) return errorTail(raw);
  if (installed === 0 && removed === 0) return ultra ? '✓' : '✓ up-to-date';
  if (ultra) return `✓+${installed}-${removed}`;
  return `✓ synced: +${installed} / -${removed} packages`;
}

function summarizeList(raw: string, ultra: boolean, noun: string): string {
  const lines = stripAnsi(raw).split('\n').filter((l) => l.trim() && !/^-+\s|^Package\s/i.test(l));
  if (ultra) return `${lines.length} ${noun}`;
  return `${lines.length} ${noun} installed\n${lines.slice(0, 25).join('\n')}`;
}

function errorTail(raw: string): string {
  const lines = stripAnsi(raw).split('\n').filter((l) => l.trim());
  const errs = lines.filter((l) => /error|failed|cannot|not found/i.test(l));
  return (errs.length ? errs : lines.slice(-8)).slice(0, 8).join('\n');
}

function result0(filtered: string, result: { exitCode: number; execMs: number }, cmdType: string, raw: string): HandlerResult {
  return {
    filteredOutput: filtered || (result.exitCode === 0 ? 'ok' : raw),
    exitCode: result.exitCode,
    rawOutput: raw,
    cmdType,
    execMs: result.execMs,
  };
}

File: internal/handlers/pkg.go
package handlers

import (
	"fmt"
	"regexp"
	"strings"

	"github.com/Kevalgor12/tok-proxy/internal/run"
	"github.com/Kevalgor12/tok-proxy/internal/util"
)

// Non-Node package managers and codegen: pip, uv, bundle, prisma, gem. Installs
// collapse to "N installed" plus any errors; listings to counts.

func HandlePip(args []string, ultra bool) Result {
	sub := ""
	if len(args) > 0 {
		sub = args[0]
	}
	r := run.Run("pip", args)
	raw := combined(r)
	var filtered string
	switch sub {
	case "install":
		filtered = summarizePipInstall(raw, ultra)
	case "list", "freeze":
		filtered = summarizeList(raw, ultra, "packages")
	default:
		filtered = strings.TrimSpace(util.StripAnsi(raw))
	}
	return finalizeRaw(filtered, r, raw, "pip "+firstNonEmpty(sub, "cmd"))
}

func HandleUv(args []string, ultra bool) Result {
	sub := ""
	if len(args) > 0 {
		sub = args[0]
	}
	r := run.Run("uv", args)
	raw := combined(r)
	var filtered string
	switch sub {
	case "sync", "pip", "add", "install":
		filtered = summarizeUvSync(raw, ultra, r.ExitCode)
	default:
		filtered = strings.TrimSpace(util.StripAnsi(raw))
	}
	return finalizeRaw(filtered, r, raw, "uv "+firstNonEmpty(sub, "cmd"))
}

var (
	bundleDoneRe  = regexp.MustCompile(`Bundle complete!\s+(\d+)\s+Gemfile dependencies,\s+(\d+)\s+gems now installed`)
	prismaGenRe   = regexp.MustCompile(`Generated Prisma Client\s+\(([^)]+)\)`)
	prismaMigRe   = regexp.MustCompile(`(?i)(Your database is now in sync|already in sync|migration.+applied)`)
	gemInstallRe  = regexp.MustCompile(`(\d+)\s+gems? installed`)
	pipErrorRe    = regexp.MustCompile(`^ERROR:`)
	pipSuccessRe  = regexp.MustCompile(`Successfully installed\s+(.+)`)
	pipSatisfyRe  = regexp.MustCompile(`(?i)already satisfied`)
	uvAddedRe     = regexp.MustCompile(`(?m)^\s*[+]\s+\S+`)
	uvRemovedRe   = regexp.MustCompile(`(?m)^\s*[-]\s+\S+`)
	listSkipRe    = regexp.MustCompile(`(?i)^-+\s|^Package\s`)
	tailErrRe     = regexp.MustCompile(`(?i)error|failed|cannot|not found`)
	whitespaceSep = regexp.MustCompile(`\s+`)
)

func HandleBundle(args []string, ultra bool) Result {
	sub := "install"
	if len(args) > 0 {
		sub = args[0]
	}
	r := run.Run("bundle", args)
	raw := combined(r)
	var filtered string
	if m := bundleDoneRe.FindStringSubmatch(util.StripAnsi(raw)); m != nil {
		if ultra {
			filtered = "✓" + m[2] + "gems"
		} else {
			filtered = fmt.Sprintf("✓ Bundle complete: %s gems (%s deps)", m[2], m[1])
		}
	} else if r.ExitCode != 0 {
		filtered = errorTail(raw)
	} else {
		filtered = strings.TrimSpace(util.StripAnsi(raw))
	}
	return finalizeRaw(filtered, r, raw, "bundle "+sub)
}

func HandlePrisma(args []string, ultra bool) Result {
	sub := ""
	if len(args) > 0 {
		sub = args[0]
	}
	r := run.Run("prisma", args)
	raw := combined(r)
	clean := util.StripAnsi(raw)
	var filtered string
	switch {
	case prismaGenRe.MatchString(clean):
		if ultra {
			filtered = "✓gen"
		} else {
			filtered = "✓ Generated Prisma Client (" + prismaGenRe.FindStringSubmatch(clean)[1] + ")"
		}
	case prismaMigRe.MatchString(clean):
		if ultra {
			filtered = "✓migrate"
		} else {
			filtered = "✓ " + prismaMigRe.FindStringSubmatch(clean)[1]
		}
	case r.ExitCode != 0:
		filtered = errorTail(raw)
	default:
		filtered = strings.TrimSpace(clean)
	}
	return finalizeRaw(filtered, r, raw, "prisma "+firstNonEmpty(sub, "cmd"))
}

func HandleGem(args []string, ultra bool) Result {
	sub := ""
	if len(args) > 0 {
		sub = args[0]
	}
	r := run.Run("gem", args)
	raw := combined(r)
	var filtered string
	switch {
	case gemInstallRe.MatchString(util.StripAnsi(raw)):
		n := gemInstallRe.FindStringSubmatch(util.StripAnsi(raw))[1]
		if ultra {
			filtered = "✓" + n + "gems"
		} else {
			filtered = "✓ " + n + " gems installed"
		}
	case sub == "list":
		filtered = summarizeList(raw, ultra, "gems")
	case r.ExitCode != 0:
		filtered = errorTail(raw)
	default:
		filtered = strings.TrimSpace(util.StripAnsi(raw))
	}
	return finalizeRaw(filtered, r, raw, "gem "+firstNonEmpty(sub, "cmd"))
}

func summarizePipInstall(raw string, ultra bool) string {
	clean := util.StripAnsi(raw)
	var errors []string
	for _, l := range strings.Split(clean, "\n") {
		if pipErrorRe.MatchString(l) {
			errors = append(errors, l)
		}
	}
	if len(errors) > 0 {
		if ultra {
			return fmt.Sprintf("✗%derr", len(errors))
		}
		return "✗ Install failed\n" + strings.Join(limit(errors, 5), "\n")
	}
	if m := pipSuccessRe.FindStringSubmatch(clean); m != nil {
		pkgs := whitespaceSep.Split(strings.TrimSpace(m[1]), -1)
		if ultra {
			return fmt.Sprintf("✓%dpkg", len(pkgs))
		}
		return fmt.Sprintf("✓ Installed %d %s: %s",
			len(pkgs), plural(len(pkgs), "package", "packages"), strings.Join(limit(pkgs, 10), ", "))
	}
	if pipSatisfyRe.MatchString(clean) {
		if ultra {
			return "✓cached"
		}
		return "✓ requirements already satisfied"
	}
	if ultra {
		return "✓"
	}
	return "✓ ok"
}

func summarizeUvSync(raw string, ultra bool, exitCode int) string {
	clean := util.StripAnsi(raw)
	installed := len(uvAddedRe.FindAllString(clean, -1))
	removed := len(uvRemovedRe.FindAllString(clean, -1))
	if exitCode != 0 {
		return errorTail(raw)
	}
	if installed == 0 && removed == 0 {
		if ultra {
			return "✓"
		}
		return "✓ up-to-date"
	}
	if ultra {
		return fmt.Sprintf("✓+%d-%d", installed, removed)
	}
	return fmt.Sprintf("✓ synced: +%d / -%d packages", installed, removed)
}

func summarizeList(raw string, ultra bool, noun string) string {
	var lines []string
	for _, l := range strings.Split(util.StripAnsi(raw), "\n") {
		if strings.TrimSpace(l) != "" && !listSkipRe.MatchString(l) {
			lines = append(lines, l)
		}
	}
	if ultra {
		return fmt.Sprintf("%d %s", len(lines), noun)
	}
	return fmt.Sprintf("%d %s installed\n%s", len(lines), noun, strings.Join(limit(lines, 25), "\n"))
}

func errorTail(raw string) string {
	lines := nonEmptyLines(util.StripAnsi(raw))
	var errs []string
	for _, l := range lines {
		if tailErrRe.MatchString(l) {
			errs = append(errs, l)
		}
	}
	if len(errs) > 0 {
		return strings.Join(limit(errs, 8), "\n")
	}
	return strings.Join(lastN(lines, 8), "\n")
}
